src/ngraph/runtime/gpu/cudnn_emitter.cpp

```diff
@@ -343,3 +343,144 @@ size_t runtime::gpu::CUDNNEmitter::build_pooling(const GPURuntimeContext* ctx,
     m_primitive_emitter->cache(hash, primitive_index);
     return primitive_index;
 }
+
+size_t runtime::gpu::CUDNNEmitter::build_batchnorm(const runtime::gpu::GPURuntimeContext* ctx,
+                                                   const cudnnBatchNormMode_t& bn_op,
+                                                   const Prop& direction,
+                                                   const Shape& tensor_shape,
+                                                   const Shape& param_shape,
+                                                   double epsilon)
+{
+    // Assumes NC{d1...dN} format
+    std::stringstream ss;
+    ss.precision(std::numeric_limits<double>::digits10 + 2);
+
+    ss << "bn_op" << bn_op << "_dir" << static_cast<int>(direction) << "_ts"
+       << join(tensor_shape, "_") << "_ps" << join(param_shape, "_") << "_eps" << epsilon;
+    std::string hash = ss.str();
+    std::replace(hash.begin(), hash.end(), '.', '_');
+
+    size_t primitive_index = m_primitive_emitter->lookup(hash);
+    if (primitive_index != std::numeric_limits<size_t>::max())
+    {
+        return primitive_index;
+    }
+
+    if (epsilon < CUDNN_BN_MIN_EPSILON)
+    {
+        throw std::runtime_error("Batch Norm epsilon is less than CUDNN_BN_MIN_EPSILON");
+    }
+
+    cudnnTensorDescriptor_t derived_param_desc;
+    cudnnCreateTensorDescriptor(&derived_param_desc);
+    auto tensor_desc = runtime::gpu::cudnn_util::tensor_descriptor_from_shape(tensor_shape);
+    cudnnDeriveBNTensorDescriptor(derived_param_desc, tensor_desc, bn_op);
+
+    float alpha = 1.0, beta = 0.0;
+    gpu::primitive* batchnorm;
```

Comment:
Possible to use shared_ptr/unique_ptr?

Reply:
I'm almost OK with using naked pointers if you do a new which is followed fairly quicly by a delete. I don't even see how this is deleted. Use *_ptr for this as @fengleitian suggests

Reply:
Good suggestions, updated in 582abbadddf9b9bed9e115ea749b3323c9f1dfda.

Previously, allocation and deletion of primitives were occuring in seperate
translation units with raw c pointers. Because of this, it was not
clear that these were being freed appropriately, nor did it indicate
ownership of the pointers.

In the above commit these raw pointers have been converted over to
std::unique_ptr so that the construction/destruction is managed
automatically. Furthermore, GPUPrimitiveEmitter::insert now only
takes an r-value reference, requiring move-semantics to indicate
that when inserting a primitive, the GPUPrimitiveEmitter takes
ownership of the pointer.

All instances of primitive creation have been modified.